src/routes.ts

```diff
@@ -53,99 +47,161 @@ async function getSessionAgent(
   }
 }
 
-export const createRouter = (ctx: AppContext) => {
-  const router = express.Router()
+export function createRouter(ctx: AppContext): RequestListener {
+  const router = express()
 
   // Static assets
   router.use('/public', express.static(path.join(__dirname, 'pages', 'public')))
 
   // OAuth metadata
   router.get(
-    '/client-metadata.json',
-    handler((_req, res) => {
-      return res.json(ctx.oauthClient.clientMetadata)
-    })
+    '/oauth-client-metadata.json',
+    handler((req: Request, res: Response) => {
+      res.json(ctx.oauthClient.clientMetadata)
+    }),
+  )
+
+  // Public keys
+  router.get(
+    '/.well-known/jwks.json',
+    handler((req: Request, res: Response) => {
+      res.json(ctx.oauthClient.jwks)
+    }),
   )
 
   // OAuth callback to complete session creation
   router.get(
     '/oauth/callback',
-    handler(async (req, res) => {
+    handler(async (req: Request, res: Response) => {
       const params = new URLSearchParams(req.originalUrl.split('?')[1])
       try {
-        const { session } = await ctx.oauthClient.callback(params)
-        const clientSession = await getIronSession<Session>(req, res, {
+        // Load the session cookie
+        const session = await getIronSession<Session>(req, res, {
           cookieName: 'sid',
           password: env.COOKIE_SECRET,
         })
-        assert(!clientSession.did, 'session already exists')
-        clientSession.did = session.did
-        await clientSession.save()
+
+        // If the user is already signed in, destroy the old credentials
+        if (session.did) {
+          try {
+            const oauthSession = await ctx.oauthClient.restore(session.did)
+            if (oauthSession) oauthSession.signOut()
+          } catch (err) {
+            ctx.logger.warn({ err }, 'oauth restore failed')
+          }
+        }
+
+        // Complete the OAuth flow
+        const oauth = await ctx.oauthClient.callback(params)
+
+        // Update the session cookie
+        session.did = oauth.session.did
+        await session.save()
       } catch (err) {
         ctx.logger.error({ err }, 'oauth callback failed')
         return res.redirect('/?error')
       }
       return res.redirect('/')
-    })
+    }),
   )
 
   // Login page
   router.get(
     '/login',
-    handler(async (_req, res) => {
-      return res.type('html').send(page(login({})))
-    })
+    handler(async (req: Request, res: Response) => {
+      res.type('html').send(page(login({})))
+    }),
   )
 
   // Login handler
   router.post(
     '/login',
-    handler(async (req, res) => {
+    express.urlencoded({ extended: true }),
```

Comment:
I don't think `extended` is necessary here.
```suggestion
    express.urlencoded(),
```